system-tests-new/fortran-prebuild/test_fortran_prebuild.py

```diff
@@ -0,0 +1,84 @@
+import logging
+import zlib
+from pathlib import Path
+from unittest import mock
+
+from fab.util import file_walk, run_command
+
+from fab.build_config import BuildConfig
+from fab.steps.analyse import Analyse
+from fab.steps.compile_fortran import CompileFortran
+from fab.steps.grab import GrabFolder, GrabPreBuild
+from fab.steps.link_exe import LinkExe
+from fab.steps.preprocess import fortran_preprocessor
+from fab.steps.walk_source import FindSourceFiles
+
+
+class TestFortranPrebuild(object):
+
+    def build_config(self, fab_workspace, prebuild_folder=None):
+        logging.getLogger('fab').setLevel(logging.WARNING)
+
+        build_config = BuildConfig(
+            project_label='test_prebuild',
+            fab_workspace=fab_workspace,
+            steps=[
+                GrabFolder(Path(__file__).parent / 'project-source', dst='src'),
+                *([GrabPreBuild(prebuild_folder)] if prebuild_folder else []),
+                FindSourceFiles(),
+                fortran_preprocessor(preprocessor='cpp -traditional-cpp -P'),
+                Analyse(root_symbol='my_prog'),
+                CompileFortran(compiler='gfortran -c', common_flags=['-J', '$output']),
+                LinkExe(flags=['-lgfortran']),
+            ],
+            multiprocessing=False,
+        )
+
+        return build_config
+
+    def test_reliable_hashes(self, tmp_path):
+        # Make sure we get the SAME MODULE HASHES in DIFFERENT FOLDERS.
+        # This is achieved by changing to the source file's folder during compilation.
+        # If we don't do this, then modules built in different workspaces can have different folders embedded in them,
+        # thereby changing their hash, thereby changing the hash of things which depend on them...
+        # build the project in "some other fab workspace", from where we'll share its prebuild
+
+        config1 = self.build_config(fab_workspace=tmp_path / 'workspace_one')
+        config1.run()
+
+        files1 = set(file_walk(config1.build_output))
+        hashes1 = {f.relative_to(config1.build_output): zlib.crc32(open(f, 'rb').read()) for f in files1}
+
+        config2 = self.build_config(fab_workspace=tmp_path / 'workspace_two')
+        config2.run()
+
+        files2 = set(file_walk(config2.build_output))
+        hashes2 = {f.relative_to(config2.build_output): zlib.crc32(open(f, 'rb').read()) for f in files2}
+
+        del hashes1[Path('__analysis.csv')]
+        del hashes2[Path('__analysis.csv')]
+        assert hashes1 == hashes2
```

Comment:
I may be misunderstanding what is being tested but this seems to be testing that the output files have the same contents. Don't you want to test that their filenames contain the same hashes?

Reply:
The main point of that test is to ensure that two identical _my_mod.f90_ produce mods with identical _file contents_, despite being in different folders. It's testing the fix for removing folders from mod contents. Of course it follows then that the two identical _my_prog.f90_ which use those mods must become artefacts with identical hash filenames, as they depend on the same mod content.

Did that make sense? I think the code is simple but the use cases can be complicated!

Reply:
(also, it's testing both content and filenames are identical, for the reasons "described")